### PR TITLE
Opt-in optimization for diffRelayQuery

### PR DESCRIPTION
**Pull request by kassens**
For the diffing step we currently traverse down every single edge and their full child tree to find the missing fields. This adds an optimization to return early in the diff process.

### CODE REVIEW

**Comment by kassens** [comment]
NOTE: this PR currently contains some changes to the Todo MVC app that I'm not planning to ship and will remove before shipping.

File: src/store/RelayFragmentTracker.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ *
+ * @providesModule RelayFragmentTracker
+ * @flow
+ */
+
+'use strict';
+
+const store: Map<string, Set<string>> = new Map();
+
+const RelayFragmentTracker = {
```
Comment (josephsavona) [inline src/store/RelayFragmentTracker.js]: let's make this a class

File: src/store/RelayFragmentTracker.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  track(clientID: string, fragmentHash: string): void {
```
Comment (josephsavona) [inline src/store/RelayFragmentTracker.js]: `dataID: DataID, ...`

File: src/query/RelayQuery.js [hunk 3]
```diff
@@ -857,11 +857,11 @@ class RelayQueryFragment extends RelayQueryNode {
     let compositeHash = this.__compositeHash__;
     if (!compositeHash) {
       // TODO: Simplify this hash function, #9599170.
-      compositeHash = generateRQLFieldAlias(
+      compositeHash = 'comp_hash_of<' + (
```
Comment (josephsavona) [inline src/query/RelayQuery.js]: :-)

File: scripts/babel-relay-plugin/src/__fixtures__/connectionWithSelectVariables.fixture [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+            ... on UserConnectionEdge @relay(selectVariables: []) {
+              node {
+                name
+              }
+            }
+          }
+        }
+        friends2: friends(first: $limit) {
+          edges {
+            ... on UserConnectionEdge @relay(selectVariables: ["size"]) {
```
Comment (josephsavona) [inline scripts/babel-relay-plugin/src/__fixtures__/connectionWithSelectVariables.fixture]: could we allow `.. on Type @relay(variables: {size: $size})` such that the developer gets complete control?

**Comment by josephsavona** [comment]
nice :-)

**Comment by kassens** [comment]
@facebook-github-bot import

**Pull request closed without merging**